go/vt/vttablet/tabletserver/schema/historian.go

```diff
@@ -0,0 +1,274 @@
+/*
+Copyright 2020 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schema
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+
+	"vitess.io/vitess/go/mysql"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/log"
+	binlogdatapb "vitess.io/vitess/go/vt/proto/binlogdata"
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	"vitess.io/vitess/go/vt/vtgate/evalengine"
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/connpool"
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/tabletenv"
+
+	"vitess.io/vitess/go/vt/sqlparser"
+)
+
+const getSchemaVersions = "select id, pos, ddl, time_updated, schemax from _vt.schema_version where id > %d order by id asc"
+
+// Historian defines the interface to reload a db schema or get the schema of a table for a given position
+type Historian interface {
+	RegisterVersionEvent() error
+	Reload(ctx context.Context) error
+	GetTableForPos(tableName sqlparser.TableIdent, pos string) *binlogdatapb.MinimalTable
+	Open() error
+	Init(tabletType topodatapb.TabletType) error
+	SetTrackSchemaVersions(val bool)
+}
+
+// HistoryEngine exposes only those schema.Engine APIs that are needed by the HistorianSvc
+// We define an interface to facilitate a mock for testing
+var _ HistoryEngine = (*Engine)(nil)
+
+// HistoryEngine defines the interface which is implemented by schema.Engine for the historian to get current schema
+// information from the database
+type HistoryEngine interface {
+	Open() error
+	Reload(ctx context.Context) error
+	GetConnection(ctx context.Context) (*connpool.DBConn, error)
+	GetSchema() map[string]*Table
+}
+
+// TrackedSchema has the snapshot of the table at a given pos (reached by ddl)
+type TrackedSchema struct {
+	schema map[string]*binlogdatapb.MinimalTable
+	pos    mysql.Position
+	ddl    string
+}
+
+var _ Historian = (*HistorianSvc)(nil)
+
+// HistorianSvc implements the Historian interface by calling schema.Engine for the underlying schema
+// and supplying a schema for a specific version by loading the cached values from the schema_version table
+// The schema version table is populated by the Tracker
+type HistorianSvc struct {
+	he                  HistoryEngine
+	lastID              int64
+	schemas             []*TrackedSchema
+	isMaster            bool
```

Comment:
Since Historian always runs, I don't think it depends on `isMaster`. Also, we should tie its life cycle with the schema engine. I think we can avoid the need for a separate HistorianEngine.

This gives me an idea that the `Historian` interface can just be part of the schema engine. We should brainstorm this out.